test/core/tsi/spiffe_ssl_transport_security_test.cc

```diff
@@ -0,0 +1,438 @@
+#endif
+      tsi_peer peer;
+      if (expect_client_success) {
+        EXPECT_EQ(
+            tsi_handshaker_result_extract_peer(base_.client_result, &peer),
+            TSI_OK);
+        tsi_peer_destruct(&peer);
+      } else {
+        EXPECT_EQ(base_.client_result, nullptr);
+      }
+      if (expect_server_success) {
+        EXPECT_EQ(
+            tsi_handshaker_result_extract_peer(base_.server_result, &peer),
+            TSI_OK);
+        tsi_peer_destruct(&peer);
+      } else {
+        EXPECT_EQ(base_.server_result, nullptr);
+      }
+    }
+
+    static void Destruct(tsi_test_fixture* fixture) {
+      auto* self = reinterpret_cast<SslTsiTestFixture*>(fixture);
+      delete self;
+    }
+
+    static struct tsi_test_fixture_vtable kVtable;
+
+    tsi_test_fixture base_;
+    std::string root_cert_;
+    tsi_ssl_server_handshaker_factory* server_handshaker_factory_;
+    tsi_ssl_client_handshaker_factory* client_handshaker_factory_;
+    std::shared_ptr<RootCertInfo> server_spiffe_bundle_map_;
+    std::shared_ptr<RootCertInfo> client_spiffe_bundle_map_;
+
+    std::string server_key_;
+    std::string server_cert_;
+    std::string client_key_;
+    std::string client_cert_;
+    bool expect_server_success_;
+    bool expect_client_success_1_2_;
+    bool expect_client_success_1_3_;
+    tsi_ssl_pem_key_cert_pair* client_pem_key_cert_pairs_;
+    tsi_ssl_pem_key_cert_pair* server_pem_key_cert_pairs_;
+  };
+};
+
+struct tsi_test_fixture_vtable
+    SpiffeSslTransportSecurityTest::SslTsiTestFixture::kVtable = {
+        &SpiffeSslTransportSecurityTest::SslTsiTestFixture::SetupHandshakers,
+        &SpiffeSslTransportSecurityTest::SslTsiTestFixture::
+            CheckHandshakerPeers,
+        &SpiffeSslTransportSecurityTest::SslTsiTestFixture::Destruct};
+
+// Valid SPIFFE Bundles on both sides with the root configured for the
+// appropriate trust domain
+TEST_P(SpiffeSslTransportSecurityTest, MTLSSpiffe) {
+  auto* fixture = new SslTsiTestFixture(
+      kServerKeyPath, kServerCertPath, kClientKeyPath, kClientCertPath,
+      kServerSpiffeBundleMapPath, kClientSpiffeBundleMapPath, kNonSpiffeCAPath,
+      /*expect_server_success=*/true,
+      /*expect_client_success_1_2=*/true, /*expected_client_success_1_3=*/true);
+  fixture->Run();
+}
+
+// Valid SPIFFE Bundles on both sides with the root configured for the
+// appropriate trust domain, and a certificate chain with an intermediate CA on
+// the server side
+TEST_P(SpiffeSslTransportSecurityTest, MTLSSpiffeChain) {
+  auto* fixture = new SslTsiTestFixture(
+      kServerChainKeyPath, kServerChainCertPath, kClientKeyPath,
+      kClientCertPath, kServerSpiffeBundleMapPath, kClientSpiffeBundleMapPath,
+      kNonSpiffeCAPath,
+      /*expect_server_success=*/true,
+      /*expect_client_success_1_2=*/true,
+      /*expected_client_success_1_3=*/true);
+  fixture->Run();
+}
+
+// Just Spiffe bundles on the client side - the server side has the root
```

Comment:
Suggested edit (for consistency with the other comments above): "Valid SPIFFE bundle on the client side, but the server side has a flat list of CA certificates."